Sources/AsyncHTTPClient/NIOTransportServices/TLSConfiguration.swift

```diff
@@ -0,0 +1,127 @@
+//
+//===----------------------------------------------------------------------===//
+
+#if canImport(Network)
+
+import Foundation
+import Network
+import NIOSSL
+import NIOTransportServices
+
+internal extension TLSVersion {
+    /// return Network framework TLS protocol version
+    var nwTLSProtocolVersion: tls_protocol_version_t {
+        switch self {
+        case .tlsv1:
+            return .TLSv10
+        case .tlsv11:
+            return .TLSv11
+        case .tlsv12:
+            return .TLSv12
+        case .tlsv13:
+            return .TLSv13
+        }
+    }
+}
+
+@available (macOS 10.14, iOS 12.0, tvOS 12.0, watchOS 6.0, *)
+internal extension TLSConfiguration {
+    
+    /// Dispatch queue used by Network framework TLS to control certificate verification
+    static var tlsDispatchQueue = DispatchQueue(label: "TLSDispatch")
+
+    /// create NWProtocolTLS.Options for use with NIOTransportServices from the NIOSSL TLSConfiguration
+    ///
+    /// - Parameter queue: Dispatch queue to run `sec_protocol_options_set_verify_block` on.
+    /// - Returns: Equivalent NWProtocolTLS Options
+    func getNWProtocolTLSOptions() -> NWProtocolTLS.Options {
+        let options = NWProtocolTLS.Options()
+        
+        // minimum TLS protocol
+        if #available(macOS 10.15, iOS 13.0, tvOS 13.0, watchOS 6.0, *) {
+            sec_protocol_options_set_min_tls_protocol_version(options.securityProtocolOptions, self.minimumTLSVersion.nwTLSProtocolVersion)
+        }
+        
+        // maximum TLS protocol
+        if let maximumTLSVersion = self.maximumTLSVersion {
+            if #available(macOS 10.15, iOS 13.0, tvOS 13.0, watchOS 6.0, *) {
+                sec_protocol_options_set_max_tls_protocol_version(options.securityProtocolOptions, maximumTLSVersion.nwTLSProtocolVersion)
+            } else {
+                precondition(self.maximumTLSVersion != nil, "TLSConfiguration.maximumTLSVersion is not supported")
+            }
+        }
+
+        // application protocols
+        if self.applicationProtocols.count > 0 {
+            preconditionFailure("TLSConfiguration.applicationProtocols is not supported")
+        }
+        /*for applicationProtocol in self.applicationProtocols {
+            applicationProtocol.utf8.withContiguousStorageIfAvailable { buffer in
+                guard let opaquePointer = OpaquePointer(buffer.baseAddress) else { return }
+                let int8Pointer = UnsafePointer<Int8>(opaquePointer)
+                sec_protocol_options_add_tls_application_protocol(options.securityProtocolOptions, int8Pointer)
+            }
+        }*/
+
+        // the certificate chain
+        if self.certificateChain.count > 0 {
+            preconditionFailure("TLSConfiguration.certificateChain is not supported")
+        }
+        
+        // cipher suites
+        if self.cipherSuites.count > 0 {
+            //preconditionFailure("TLSConfiguration.cipherSuites is not supported")
```

Comment:
I did have an attempt at this, and got quite far but then hit an issue which would have taken time to get around and I wasn't sure it was worth the development time.

Reply:
apple/swift-nio-ssl#207.

Reply:
Would this use ssl_create_cipher_list(). Are the cipher ids consistent across libraries?

Reply:
It can't, that function is private. It would have to create a throwaway `SSL_CTX` and call `SSL_CTX_set_cipher_list` followed by `SSL_CTX_get_ciphers`. For BoringSSL you want to get the cipher _value_, not the cipher _id_, because the cipher value is the UInt16 value associated with that cipher in the wire protocol. That is consistent across libraries.

Reply:
I'd rather not have the commented out code: can we just add a TODO with a link to the relevant issue?